0015-authority.md

```diff
@@ -0,0 +1,478 @@
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## Scrapping per NetworkBehaviour Authority
+See appendix for details. There are very little actual use cases for this and scrapping this will more then half the code size of the authority feature and lower the bandwidth usage of MLAPI.
+
+## Naming
+We might want to consider a different name for authority. Some candidates would be `AuthorityMode`, `ControlMode` or `GhostMode`(to match DOTS). Similar `HasAuthority` might be `HasControl`.
+
+## Snapshots / AOI
+There are a lot of unknowns on how our snapshot and AOI system will work with `NetworkVariablePermission.Authority`. We do not plan to resolve this unresolved question in the authority RFC as it only introduces an additional `NetworkVariablePermission`. The snapshot / AOI RFCs will answer this question.
+
+## AuthorityChanged Event
+Do we need an AuthorityChanged event? We already have `OnAuthorityGained` and `OnAuthorityLost` events but would we need an additional to event to signal if the authority mode itself changed?
+
+## Default NetworkVariable Permissions
+
+The default write permission for a NetworkVariable is currently `NetworkVariablePermission.ServerOnly` and the default write permission is `NetworkVariablePermission.Everyone`. We could consider changing the default write permission to `NetworkVariablePermission.Authority`. It is not clear whether that would simplify or complicate the onboarding experience of MLAPI.
+
+## Shared/Distributed Authority
+
+The RFC does not cover authority modes where multiple peers have authority over an object. It also does not cover prediction. The proposed authority system can be extended to support more authority modes. To support shared/distributed authority we might need an ACL entry to indicate which peers have authority over an object
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Prediction
+The authority feature is a first step towards the goal of supporting client side prediction. When using client side prediction we need precises control over what's being run on the predicted client and run the same on the server. The `HasAuthority` flag will allow us to do this by setting it to true on both the predicting client and the server.
+
+## Commands
+Commands in multiplayer games are inputs which are created by the owner/controller of an object and then sent to the authority for execution. The authority mode feature will allow us to implement commands in the future.
+
+## RPCs
+
+Right now RPCs can only be send to the server/clients. This means that a lot of time boilerplate must be introduced to write authoritative concepts.
+```csharp
+if(HasAuthority)
+{
+    SetPosition(m_TargetPosition);
+}
+else
+{
+    SetPositionServerRpc(m_TargetPosition);
+}
+```
+
+In the future we could consider introducing an `AuthorityRpc` or similar which sends an RPC to the authority.
```

Comment:
Something interesting with this is you wouldn't have to care about whether you're sending an RPC to a server or a client. If I have a client authority, that RPC would use the current object's owner to know to which client it should send the RPC.
I feel like having that concept of authority wouldn't be complete without RPCs. You cover netvars only here, but from a user's point of view, I would expect Authority to work with both Netvars and RPCs, no?